stdlib/public/core/IntegerParsing.swift

```diff
@@ -146,31 +45,44 @@ extension FixedWidthInteger {
   ///     `radix`.
   ///   - radix: The radix, or base, to use for converting `text` to an integer
   ///     value. `radix` must be in the range `2...36`. The default is 10.
-  @inlinable // @specializable
-  @_semantics("optimize.sil.specialize.generic.partial.never")
+  @inlinable @inline(__always)
   public init?<S: StringProtocol>(_ text: S, radix: Int = 10) {
-    _precondition(2...36 ~= radix, "Radix not in range 2...36")
-
-    if let str = text as? String, str._guts.isFastUTF8 {
-      guard let ret = str._guts.withFastUTF8 ({ utf8 -> Self? in
-        var iter = utf8.makeIterator()
-        return _parseASCII(codeUnits: &iter, radix: Self(radix))
-      }) else {
-        return nil
-      }
-      self = ret
-      return
+    precondition(2...36 ~= radix)
+    // Check we can load the UTF-8 in UInt64 chunks.
+    let wholeGuts = text._wholeGuts
+    let canChunkLoad: Bool
+    if S.self == String.self {
+      canChunkLoad = wholeGuts.isSmall || wholeGuts.count >= 9 // +1 for sign.
+    } else {
+      canChunkLoad = (wholeGuts.isSmall ? 16 : wholeGuts.count)
+        &- text.startIndex._encodedOffset >= 9
+    }
+    if _fastPath(canChunkLoad), _fastPath(wholeGuts.isFastUTF8),
```

Comment:
I think there's definitely a good case for removing the specialization for the default case here, as that one is probably not worth it – even though it would mean uncommon radix performance would be regressing.

We need the separate generic implementation not just because of cases lacking fast UTF-8 but also for those that don't fit into a UInt64.

In terms of forcing UTF-8, that might be worth it, and maybe even copying in the less-than 9 char non-smol case. So what about this line-up:
- Base 10/2/16, type fits UInt64: force UTF-8 (and smol if count < 9) and use a radix specialized path
- Uncommon radix, or type > UInt64: generic slow path

Reply:
Right, a generic unspecialized fall-back path would be slow. I'm not too concerned whether that path forces contiguous UTF-8, and it seems that instead just yielding an iterator is sufficient. The specialized fast paths might as well force contiguous UTF-8 IMO.